mat/cdense.go

```diff
@@ -57,3 +59,91 @@ func NewCDense(r, c int, data []complex128) *CDense {
 		capCols: c,
 	}
 }
+
+// reuseAs resizes an empty matrix to a r×c matrix,
+// or checks that a non-empty matrix is r×c.
+//
+// reuseAs must be kept in sync with reuseAsZeroed.
+func (m *CDense) reuseAs(r, c int) {
+	if m.mat.Rows > m.capRows || m.mat.Cols > m.capCols {
+		// Panic as a string, not a mat.Error.
+		panic("mat: caps not correctly set")
+	}
+	if r == 0 || c == 0 {
+		panic(ErrZeroLength)
+	}
+	if m.IsZero() {
+		m.mat = cblas128.General{
+			Rows:   r,
+			Cols:   c,
+			Stride: c,
+			Data:   useC(m.mat.Data, r*c),
+		}
+		m.capRows = r
+		m.capCols = c
+		return
+	}
+	if r != m.mat.Rows || c != m.mat.Cols {
+		panic(ErrShape)
+	}
+}
+
+func (m *CDense) reuseAsZeroed(r, c int) {
+	m.reuseAs(r, c)
```

Comment:
In Dense (and here currently), it says we must be kept in sync, but I don't see any reason why this kind of implementation can't work. Do you know why @kortschak ?

Reply:
This does extra work. I imagine that's why I had two separate routines.

Reply:
Having two routines makes sense, it's that the Dense code has the reuseAs code copied rather than called as a method

Reply:
That's because the `use` is hardcoded into `reuseAs` and `useZeroed` in `reuseAsZeroed`. If you do `reuseAs` and then `zero`, you may allocate and then re-zero the data field.

Reply:
Ahh, yes.